raw_decoder/raw_decoder.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	commonData := &CommonData{
+		Level: 30,
+	}
+
+	baseProto := Proto{
+		CommonData: commonData,
+	}
+
+	if v := raw["have_ar"]; v != nil {
+		res, ok := v.(bool)
+		if ok {
+			baseProto.HaveAr = &res
+		}
+	}
+	if v := raw["uuid"]; v != nil {
+		baseProto.Uuid, _ = v.(string)
+	}
+	if v := raw["username"]; v != nil {
+		baseProto.Account, _ = v.(string)
+	}
+	if v := raw["trainerlvl"]; v != nil {
+		lvl, ok := v.(float64)
+		if ok {
+			baseProto.Level = int(lvl)
+		}
+	}
+	if v := raw["scan_context"]; v != nil {
+		baseProto.ScanContext, _ = v.(string)
+	}
+	if v := raw["lat_target"]; v != nil {
+		baseProto.Lat, _ = v.(float64)
+	}
+	if v := raw["lon_target"]; v != nil {
+		baseProto.Lon, _ = v.(float64)
+	}
+
+	contents, ok := raw["contents"].([]any)
+	if !ok {
+		return nil, errors.New("failed to decode 'contents'")
+	}
+
+	var protos []Proto
+
+	for _, v := range contents {
+		entry, ok := v.(map[string]any)
+		if !ok {
+			continue
+		}
+		// Try to decode the payload automatically without requiring any knowledge of the
+		// provider type
+
+		base64data := decodeAlternate(entry, "data", "payload")
+		method := decodeAlternate(entry, "method", "type")
+		if method == nil || base64data == nil {
+			log.Errorf("Error decoding raw (no method or base64data)")
+			continue
+		}
+
+		proto := baseProto
+		proto.base64Response, _ = base64data.(string)
+		proto.Method = func() int {
+			if res, ok := method.(float64); ok {
+				return int(res)
+			}
+			return 0
+		}()
+
+		if request := entry["request"]; request != nil {
+			proto.base64Request, _ = request.(string)
+		}
+		if haveAr := entry["have_ar"]; haveAr != nil {
+			res, ok := haveAr.(bool)
+			if ok {
+				proto.HaveAr = &res
+			}
+		}
+
+		protos = append(protos, proto)
+	}
+	return &ProtoData{
+		CommonData: commonData,
+		Protos:     protos,
+	}, nil
+}
+
+func (dec *rawDecoder) GetProtoDataFromGRPC(in *grpc.RawProtoRequest) *ProtoData {
+	commonData := &CommonData{
+		Uuid:    in.DeviceId,
+		Account: in.Username,
+		Level:   int(in.TrainerLevel),
+	}
+
+	baseProto := Proto{
+		CommonData: commonData,
+		Lat:        float64(in.LatTarget),
+		Lon:        float64(in.LonTarget),
+		HaveAr:     in.HaveAr,
+	}
+
+	if in.ScanContext != nil {
+		baseProto.ScanContext = *in.ScanContext
+	}
+
+	protos := make([]Proto, len(in.Contents))
+	for i, v := range in.Contents {
+		proto := baseProto
+		proto.Method = int(v.Method)
+		proto.requestBytes = v.RequestPayload
+		proto.responseBytes = v.ResponsePayload
+		if v.HaveAr != nil {
+			proto.HaveAr = v.HaveAr
+		}
+		protos[i] = proto
+	}
+	return &ProtoData{
+		CommonData: commonData,
+		Protos:     protos,
+	}
+}
+
+func (dec *rawDecoder) Decode(ctx context.Context, protoData *ProtoData, decodeFn func(context.Context, *Proto)) {
+	for _, proto := range protoData.Protos {
+		// provide independent cancellation contexts for each proto decode
+		ctx, cancel := context.WithTimeout(ctx, dec.decodeTimeout)
+		decodeFn(ctx, &proto)
```

Comment:
We used to call each decoder in their own go-routine; however this caused too much tripping up and waiting on the striped mutex - since you might have responses that affect the same objects (eg during questing etc).
You also see some MITM include the gmo and the encounters in the same raw (even though this can't be how they were executed against the game).
Having independent cancellation is fine; originally intention was to be quite strict to not end in a backlog situation (which RDM is very common for when overloaded) - where you never catch up because you have so many concurrent decodes happening.

Reply:
This is actually copy/paste from existing code. But It looks like we don't need to make a new context for every single call here.

Reply:
(I'll reply to the other comments when I can )